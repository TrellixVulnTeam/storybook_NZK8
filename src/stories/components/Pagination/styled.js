// @flow

import styled from 'styled-components';
import { colors } from 'utils/StyleGuide';

export const PageWrapper = styled.li`
  background-color: ${p => (p.isActive ? colors.brightBlue : colors.white)};
  color: ${colors.paleGrey};
  font-size: 14px;
  font-weight: 600;
  height: 35px;
  padding: 8px 15px;
  text-align: center;
  border: 1px solid #ccc;
  margin: 0px;
  /* background-clip: padding-box; */
  span {
    color: ${p => (p.isActive ? colors.white : colors.slateGrey)};
    cursor: pointer;
  }
`;

export const Wrapper = styled.ul`
  margin: 0px;
  padding: 0px;
  list-style-type: none;
  display: inline-flex;
  flex-direction: ${p => (p.verticalAlign ? 'column' : 'row')};
  @media (max-width: 768px) {
    flex-direction: column;
  }

  li:first-child {
    left-border-radius: 3px;
  }

  li:last-child {
    right-border-radius: 3px;
  }
`;
