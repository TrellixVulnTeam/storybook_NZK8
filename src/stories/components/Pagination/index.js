// @flow

import React, { PureComponent } from 'react';
import BasePaginator from 'paginator';
import Page from './page';
import { Wrapper } from './styled';

type Props = {
  totalItemsCount: number,
  onChange: Function,
  activePage?: number,
  itemsCountPerPage?: number,
  pageRangeDisplayed?: number,
  prevPageText?: string,
  nextPageText?: string,
  lastPageText?: string,
  firstPageText?: string,
  hideFirstLastPages?: boolean,
  hideDisabled?: boolean,
  hideNavigation?: boolean,
  verticalAlign?: boolean,
};

class Pagination extends PureComponent<Props> {
  // eslint-disable-next-line react/static-property-placement
  static defaultProps = {
    itemsCountPerPage: 10,
    pageRangeDisplayed: 5,
    activePage: 1,
    prevPageText: '⟨',
    firstPageText: '«',
    nextPageText: '⟩',
    lastPageText: '»',
    hideFirstLastPages: false,
    verticalAlign: false,
  };

  isFirstPageVisible = (hasPreviousPage: boolean) => {
    const { hideDisabled, hideFirstLastPages } = this.props;
    if (hideFirstLastPages || (hideDisabled && !hasPreviousPage)) {
      return false;
    }
    return true;
  };

  isPrevPageVisible = (hasPreviousPage: boolean) => {
    const { hideDisabled, hideNavigation } = this.props;
    if (hideNavigation || (hideDisabled && !hasPreviousPage)) {
      return false;
    }
    return true;
  };

  isNextPageVisible = (hasNextPage: boolean) => {
    const { hideDisabled, hideNavigation } = this.props;
    if (hideNavigation || (hideDisabled && !hasNextPage)) {
      return false;
    }
    return true;
  };

  isLastPageVisible = (hasNextPage: boolean) => {
    const { hideDisabled, hideFirstLastPages } = this.props;
    if (hideFirstLastPages || (hideDisabled && !hasNextPage)) {
      return false;
    }
    return true;
  };

  /**
   * create pages from page component
   * @return {React$Node}
   */
  buildPages() {
    const pages = [];
    const {
      itemsCountPerPage,
      pageRangeDisplayed,
      activePage,
      prevPageText,
      nextPageText,
      firstPageText,
      lastPageText,
      totalItemsCount,
      onChange,
    } = this.props;

    const paginationInfo = new BasePaginator(itemsCountPerPage, pageRangeDisplayed).build(
      totalItemsCount,
      activePage,
    );

    for (let i = paginationInfo.first_page; i <= paginationInfo.last_page; i += 1) {
      pages.push(
        <Page
          isActive={i === activePage}
          key={i}
          pageNumber={i}
          pageText={`${i}`}
          onClick={onChange}
        />,
      );
    }

    // eslint-disable-next-line no-unused-expressions
    this.isPrevPageVisible(paginationInfo.has_previous_page) &&
      pages.unshift(
        <Page
          key={`prev${paginationInfo.previous_page}`}
          pageNumber={paginationInfo.previous_page}
          onClick={onChange}
          pageText={prevPageText}
          isDisabled={!paginationInfo.has_previous_page}
        />,
      );

    // eslint-disable-next-line no-unused-expressions
    this.isFirstPageVisible(paginationInfo.has_previous_page) &&
      pages.unshift(
        <Page
          key="first"
          pageNumber={1}
          onClick={onChange}
          pageText={firstPageText}
          isDisabled={!paginationInfo.has_previous_page}
        />,
      );

    // eslint-disable-next-line no-unused-expressions
    this.isNextPageVisible(paginationInfo.has_next_page) &&
      pages.unshift(
        <Page
          key={`next${paginationInfo.next_page}`}
          pageNumber={paginationInfo.next_page}
          onClick={onChange}
          pageText={nextPageText}
          isDisabled={!paginationInfo.has_next_page}
        />,
      );

    // eslint-disable-next-line no-unused-expressions
    this.isLastPageVisible(paginationInfo.has_next_page) &&
      pages.unshift(
        <Page
          key="last"
          pageNumber={paginationInfo.total_pages}
          onClick={onChange}
          pageText={lastPageText}
          isDisabled={paginationInfo.current_page === paginationInfo.total_pages}
        />,
      );

    return pages;
  }

  /**
   * short and sweet render
   * @return {React$Node}
   */
  render() {
    const pages = this.buildPages();
    return <Wrapper verticalAlign={this.props.verticalAlign}>{pages}</Wrapper>;
  }
}

export default Pagination;
