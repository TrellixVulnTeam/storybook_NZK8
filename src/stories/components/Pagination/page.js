// @flow
import React, { PureComponent } from 'react';
import { PageWrapper } from './styled';

type Props = {
  pageText?: string,
  pageNumber: number,
  onClick: Function,
  isActive?: boolean,
  isDisabled?: boolean,
};

class Page extends PureComponent<Props> {
  // eslint-disable-next-line react/static-property-placement
  static defaultProps = {
    isActive: false,
    isDisabled: false,
  };

  handleClick = (e: Object) => {
    const { isDisabled, pageNumber } = this.props;
    e.preventDefault();
    if (isDisabled) {
      return;
    }
    this.props.onClick(pageNumber);
  };

  /**
   * render a customizable page
   * @return {React$Node}
   */
  render() {
    const { pageText } = this.props;
    return (
      <PageWrapper isActive={this.props.isActive} onClick={this.handleClick}>
        <span>{pageText}</span>
      </PageWrapper>
    );
  }
}

export default Page;
