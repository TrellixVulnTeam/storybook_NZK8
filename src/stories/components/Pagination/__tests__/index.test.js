import React from 'react';
import { mount } from 'enzyme';
import Pagination from 'components/Pagination/index';

describe('<Pagination />', () => {
  const mandatoryProps = {
    totalItemsCount: 10,
    onChange: jest.fn(),
  };

  it('should correctly handle clicks with mandatory props', () => {
    const props = {
      ...mandatoryProps,
    };
    const sut = mount(<Pagination {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly with optional params', () => {
    const props = {
      ...mandatoryProps,
      activePage: 2,
      itemsCountPerPage: 5,
      pageRangeDisplayed: 5,
      prevPageText: '<-',
      nextPageText: '->',
      lastPageText: '>>',
      firstPageText: '<<',
      hideFirstLastPages: false,
      hideDisabled: true,
      hideNavigation: false,
      verticalAlign: true,
    };
    const sut = mount(<Pagination {...props} />);

    expect(sut).toMatchSnapshot();
  });
});
