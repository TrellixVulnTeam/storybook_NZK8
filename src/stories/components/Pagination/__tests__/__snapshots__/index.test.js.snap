// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should correctly handle clicks with mandatory props 1`] = `
<Pagination
  activePage={1}
  firstPageText="«"
  hideFirstLastPages={false}
  itemsCountPerPage={10}
  lastPageText="»"
  nextPageText="⟩"
  onChange={[MockFunction]}
  pageRangeDisplayed={5}
  prevPageText="⟨"
  totalItemsCount={10}
  verticalAlign={false}
>
  <styled__Wrapper
    verticalAlign={false}
  >
    <ul
      className="styled__Wrapper-dxl0zg-1 cggFIw"
    >
      <Page
        isActive={false}
        isDisabled={true}
        key="last"
        onClick={[MockFunction]}
        pageNumber={1}
        pageText="»"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              »
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={false}
        isDisabled={true}
        key="next2"
        onClick={[MockFunction]}
        pageNumber={2}
        pageText="⟩"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              ⟩
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={false}
        isDisabled={true}
        key="first"
        onClick={[MockFunction]}
        pageNumber={1}
        pageText="«"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              «
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={false}
        isDisabled={true}
        key="prev0"
        onClick={[MockFunction]}
        pageNumber={0}
        pageText="⟨"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              ⟨
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={true}
        isDisabled={false}
        key="1"
        onClick={[MockFunction]}
        pageNumber={1}
        pageText="1"
      >
        <styled__PageWrapper
          isActive={true}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 gYkfmV"
            onClick={[Function]}
          >
            <span>
              1
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
    </ul>
  </styled__Wrapper>
</Pagination>
`;

exports[`<Pagination /> should render correctly with optional params 1`] = `
<Pagination
  activePage={2}
  firstPageText="<<"
  hideDisabled={true}
  hideFirstLastPages={false}
  hideNavigation={false}
  itemsCountPerPage={5}
  lastPageText=">>"
  nextPageText="->"
  onChange={[MockFunction]}
  pageRangeDisplayed={5}
  prevPageText="<-"
  totalItemsCount={10}
  verticalAlign={true}
>
  <styled__Wrapper
    verticalAlign={true}
  >
    <ul
      className="styled__Wrapper-dxl0zg-1 dKYpTi"
    >
      <Page
        isActive={false}
        isDisabled={false}
        key="first"
        onClick={[MockFunction]}
        pageNumber={1}
        pageText="<<"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              &lt;&lt;
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={false}
        isDisabled={false}
        key="prev1"
        onClick={[MockFunction]}
        pageNumber={1}
        pageText="<-"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              &lt;-
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={false}
        isDisabled={false}
        key="1"
        onClick={[MockFunction]}
        pageNumber={1}
        pageText="1"
      >
        <styled__PageWrapper
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 fOgClc"
            onClick={[Function]}
          >
            <span>
              1
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
      <Page
        isActive={true}
        isDisabled={false}
        key="2"
        onClick={[MockFunction]}
        pageNumber={2}
        pageText="2"
      >
        <styled__PageWrapper
          isActive={true}
          onClick={[Function]}
        >
          <li
            className="styled__PageWrapper-dxl0zg-0 gYkfmV"
            onClick={[Function]}
          >
            <span>
              2
            </span>
          </li>
        </styled__PageWrapper>
      </Page>
    </ul>
  </styled__Wrapper>
</Pagination>
`;
